src/server/analytic_services/prometheus_reporting.js

```diff
@@ -2,359 +2,423 @@
+        type: 'Gauge',
+        name: 'reduction_ratio',
+        generate_default_set: true,
+        configuration: {
+            help: 'Object Efficiency Ratio'
+        }
+    }, {
+        type: 'Gauge',
+        name: 'object_savings_logical_size',
+        generate_default_set: true,
+        configuration: {
+            help: 'Object Savings Logical'
+        }
+    }, {
+        type: 'Gauge',
+        name: 'object_savings_physical_size',
+        generate_default_set: true,
+        configuration: {
+            help: 'Object Savings Physical'
+        }
+    }, {
+        type: 'Gauge',
+        name: 'rebuild_progress',
+        generate_default_set: true,
+        configuration: {
+            help: 'Rebuild Progress'
+        }
+    }, {
+        type: 'Gauge',
+        name: 'rebuild_time',
+        generate_default_set: true,
+        configuration: {
+            help: 'Rebuild Time'
+        }
+    }, {
+        type: 'Gauge',
+        name: 'bucket_status',
+        configuration: {
+            help: 'Bucket Health',
+            labelNames: ['bucket_name']
+        }
+    }, {
+        type: 'Gauge',
+        name: 'bucket_capacity',
+        configuration: {
+            help: 'Bucket Capacity Precent',
+            labelNames: ['bucket_name']
+        }
+    }, {
+        type: 'Gauge',
+        name: 'bucket_quota',
+        configuration: {
+            help: 'Bucket Quota Precent',
+            labelNames: ['bucket_name']
+        }
+    }, {
+        type: 'Gauge',
+        name: 'resource_status',
+        configuration: {
+            help: 'Resource Health',
+            labelNames: ['resource_name']
+        }
+    }, {
+        type: 'Gauge',
+        name: 'system_links',
+        configuration: {
+            help: 'System Links',
+            labelNames: ['resources', 'buckets', 'dashboard']
+        }
+    }, {
+        type: 'Gauge',
+        name: 'health_status',
+        generate_default_set: true,
+        configuration: {
+            help: 'Health status'
+        }
     }
+]);
 
-    export_metrics() {
-        const exported = _.map(METRIC_RECORDS, metric =>
-                this._prom_client.register.getSingleMetricAsString(metric.configuration.name))
-            .join('\n');
+class NooBaaCoreReport extends BasePrometheusReport {
+    constructor() {
+        super();
 
-        return exported;
+        this._metrics = {};
+        for (const m of NOOBAA_CORE_METRICS) {
+            this._metrics[m.name] = new prom_client[m.type]({
+                ...m.configuration,
+                name: `${this.metric_prefix}${m.name}`,
+                registers: [this._registry]
+            });
+
+            // Create a default setter, are moving the name into a local var
+            // in order to prevent the closure to capture the entire m object.
+            const { name } = m;
+            if (m.generate_default_set) {
+                this[`set_${m.name}`] = data => this._metrics[name].set(data);
```

Comment:
New function is added to class. Would we have performance related issue if there are many metrics? Would it be simple to just provide function in the base class e.g.? 
Also is set() only for Gauge type of metric? we should call corresponding function based upon metric type.

```
function observe(labels, data) {
   if (this._metrics[name].type === 'Counter') {
       this._metrics[name].inc(labels, data);
   }
   else if (this._metrics[name].type === 'Gauge') {
       this._metrics[name].set(labels, data);
   } else if (this._metrics[name].type === 'Histogram') {
       this._metrics[name].observe(labels, data);
   }
```

Reply:
@paulxuzhang I also don't like adding methods on the fly like that. 
It is an artifact of the old code which gathered metrics for core custom metrics. 
I didn't want to touch it in order to not break the existing metrics collection.

Regarding performance: each report is a singleton which means that only one instance of the class and each of the methods will be instantiated. And because of the way javascript works, a method on an object might be even faster than a method on the object's prototype. But in modern js engines (like V8) it does not really matter and the JIT might even inline in the call site in both cases

Reply:
I think its easier to understand, maintain, and add custom metrics when they are written in a JSON format in a human-readable manner rather than code.

Reply:
I must disagree writing it as part of the constructor like that:
```
this._metrics = {
   demo_metric: new client.Gauge({
       name: this.get_prefixed_name('demo_metric'),
       registers: [this.registry],
       help: 'Description line for demo_metric',
       // Other configurqation that should apply for this metric
       }),
 };
```
has the exact same benefits and the exact same number of lines of codes and clarity. 

And trying to generate function automatically will only make things more confusing for the reader/consumer:
- You cannot find all the setter when reading the class and even worse when some of them are also defined in the class (the more complicated ones)
- The JS linter cannot know about these methods 
- You will always have exceptions that you will implement in the class itself which means that in order to check for a method you need to understand that some of them are generated and some not. 
These exceptions just complicate things even more because  now the reader needs to check a flag on the "metadata" in order to understand if a setter is generated or not

I believe that this kind of generation just add to the code complexity and I strongly suggest not to use it in future reports

Reply:
ok